src/SUnit-Tests/ProcessMonitorTestServiceTest.class.st

```diff
@@ -0,0 +1,556 @@
+	processWithHalt := self fork: [ Processor activeProcess suspend. Halt new signal ].
+	
+	processWithHalt on: Halt do: [ :actual | ].
+	processWithHalt resume.
+	
+	self assert: executed.
+	self deny: testService shouldSuspendBackgroundFailures
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testResumeFailedProcessesWhenTestFailureIsUnhandled [
+	| executed |
+	executed := false.
+	self fork: [ testService suspendBackgroundFailure: Error new. executed := true ].
+	self deny: executed.	
+
+	testService handleException: Error new. "UnhandledError always started with original error signal"
+	testService handleException: UnhandledError new.
+	self assert: executed.
+	"Unhandled exception opens debugger and therefore any new errors should not be suspended"
+	self deny: testService shouldSuspendBackgroundFailures 
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testShouldSuspendBackgroundFailuresByDefault [
+
+	self assert: testService shouldSuspendBackgroundFailures 
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testStoreAllForkedProcesses [
+
+	| process |
+	process := self fork: [ ].
+	
+	self assert: (testService forkedProcesses includes: process)
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testSuspendBackgroundError [
+
+	| process error suspendedError |
+	error := Error new messageText: 'test error'.
+	
+	process := self fork: [ error signal ].
+	
+	self assert: process isSuspended.
+	suspendedError := testService suspendedBackgroundFailures at: process.
+	self assert: suspendedError class equals: UnhandledError. 
+	self assert: suspendedError exception identicalTo: error.
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testSuspendBackgroundWarning [
+	
+	| process warning suspendedError |
+	warning := Warning new.
+	process := self fork: [ warning signal ].
+	
+	self assert: process isSuspended.
+	suspendedError := testService suspendedBackgroundFailures at: process.
+	self assert: suspendedError class equals: UnhandledError. 
+	self assert: suspendedError exception identicalTo: warning.
+]
+
+{ #category : #tests }
+ProcessMonitorTestServiceTest >> testSuspendResumeTerminateIssue [
+	"The test discovers the issue with Process>>isTerminated method
+	when process was suspended in the middle and then resumed.
+	During the bootstrap process on CI it fails RunKernelTests stage.
+	But it is working on full image"
+	| process ended started |
+	ended := false.
+	started := false.
+	process := [ started := true. Processor activeProcess suspend. ended := true ] newProcess.
+	process priority: Processor activePriority + 1.
+	processes add: process.	
+	process resume.
+	self assert: started description: 'process should be started'.
+	self assert: process isSuspended description: 'process should be suspended'.
+
+	process resume.
+	self assert: ended description: 'process should be ended'.
+	self assert: process isTerminated description: 'process should be terminated'
```

Comment:
And this is failed assertion. As you see the last instruction in the block is executed (ended is true) but process is not terminated. 
I did more deep debugging and found that in that case process pc is zero (and isTerminated relies on greater pc number).
According to Eliot the zero pc is a sign of VM bug in the mapping code between jitted code and bytecode. But it is strange that it appears only with #suspend call at the middle of process (see next test which proves it)